feat: add ruby plugin (#1395)
canonical/craft-parts: docs/common/craft-parts/reference/plugins/ruby_plugin.rst

.. _craft_parts_ruby_plugin:

Ruby plugin
===========

The Ruby plugin can be used in projects that need to bundle a specific
version and variant of the Ruby interpreter within the final artifact. The
plugin can further be used to install individual gems or trigger a build tool
such as rake or bundler.

.. _craft_parts_ruby_plugin-keywords:

Keys
----

This plugin provides the following unique keys.

ruby-gems
~~~~~~~~~

**Type:** list of strings

> List of gems to be installed.

Reviewer: ```suggestion
The gems to install.
```

ruby-use-bundler
~~~~~~~~~~~~~~~~

**Type:** boolean

**Default:** False

When set to ``true``, the plugin runs the ``bundle`` command as part of the
``build`` step.

ruby-flavor
~~~~~~~~~~~

**Type:** string

**Default:** ruby

Specifies which implementation of the Ruby interpreter the plugin
should build. Supported values are ``ruby``, ``jruby``, ``rbx``,

> ``truffleruby``, and ``mruby``.

Reviewer: ```suggestion
The Ruby interpreter to compile and include. The supported interpreters are:

- ``ruby``
- ``jruby``
- ``rbx``
- ``truffleruby``
- ``mruby``
```

I don't believe we need to keep the word "implementation", but I'm not very familiar.

ruby-version
~~~~~~~~~~~~

**Type:** string

**Default:** "3.2"

The version of Ruby to build.

ruby-prefix
~~~~~~~~~~~

**Type:** string

**Default:** "/usr"

The root directory where the Ruby executables and libraries will be installed.

ruby-shared
~~~~~~~~~~~

**Type:** boolean

**Default:** False

When set to ``true``, the plugin builds ``libruby.so``, a shared library that
other binaries can link against.

ruby-use-jemalloc
~~~~~~~~~~~~~~~~~

**Type:** boolean

**Default:** False

When set to ``true``, the plugin compiles Ruby with support for the jemalloc

> memory allocator.

Reviewer: ```suggestion
Whether to compile Ruby with support for jemalloc.
```

ruby-configure-options
~~~~~~~~~~~~~~~~~~~~~~

**Type:** list of strings

> Extra arguments to pass to the compiler when building the Ruby interpreter.

Author: I would expect so, since each is an entirely different implementation with its own optional features.. There are some examples in the ruby-install README: https://github.com/postmodern/ruby-install/blob/master/README.md#synopsis

.. _ruby-details-begin:

Dependencies
------------

By default, this plugin downloads and runs the `ruby-install
<https://github.com/postmodern/ruby-install>` tool. If recompiling the
interpreter for every Ruby-based part is not desired, a ``ruby-deps`` part
can install Ruby and any gems in the build system. Other parts that use this

> interpreter must run ``after`` the ``ruby-deps`` part:

Reviewer: ```suggestion
<https://github.com/postmodern/ruby-install>` tool.

If a project has multiple Ruby parts, by default each compiles and bundles its
own Ruby interpreter. If a special Ruby part named ``ruby-deps`` is defined, it
creates a shared interpreter and shared gems that the other Ruby parts can use.
For a Ruby part to use these shared files, it must list ``ruby-deps`` in its
``after`` key.
```

Author: Thank you for a much better explanation. My attempt to adapt language from the poetry plugin reference did not turn out well.

Bundler, and install gems.

.. _craft_parts_ruby_plugin-keywords:

Keys
----

This plugin provides the following unique keys.

ruby-gems
~~~~~~~~~

**Type:** list of strings

The gems to install.

ruby-use-bundler
~~~~~~~~~~~~~~~~

**Type:** boolean

**Default:** False

Whether to use Bundler to build the gems.

ruby-flavor
~~~~~~~~~~~

**Type:** string

**Default:** None

The Ruby interpreter to build and include. The supported interpreters are:

- ``ruby``
- ``jruby``
- ``truffleruby``
- ``mruby``

ruby-version
~~~~~~~~~~~~

**Type:** string

**Default:** None

The version of the Ruby interpreter to build.

ruby-shared
~~~~~~~~~~~

**Type:** boolean

**Default:** False

When set to ``true``, the plugin builds ``libruby.so``, a shared library that
other binaries can link against.

ruby-use-jemalloc
~~~~~~~~~~~~~~~~~

**Type:** boolean

**Default:** False

Whether to build Ruby with support for jemalloc.

ruby-configure-options
~~~~~~~~~~~~~~~~~~~~~~

**Type:** list of strings

Extra arguments to pass to the ``configure`` script when building the Ruby
interpreter.

.. _ruby-details-begin:

Dependencies
------------

> The Ruby plugin needs the ``ruby`` interpreter to run Ruby programs but does not

Reviewer: > `ruby` interpreter

Nitpick: do we really need to specify the binary name here and in the next paragraph? Doesn't "Ruby interpreter" mean the same thing?

> provision it by default, to allow flexibility in the choice of interpreter flavor

Reviewer: ```suggestion
provide it by default, to give you flexibility in the choice of interpreter flavor
```

I think "provide" is a less expensive word for what we mean here.

Author: I agree and I'll accept this suggestion, but FWIW this language came straight from the Go plugin reference.
